common/djangoapps/third_party_auth/tests/test_decorators.py

```diff
@@ -57,3 +70,67 @@ def test_feature_flag_off(self, url):
             request = self.construct_request(url)
             response = mock_view(request)
             self.assertEqual(response['X-Frame-Options'], 'DENY')
+
+
+@unittest.skipIf(
+    'third_party_auth' not in settings.INSTALLED_APPS,
+    'third_party_auth is not currently installed in CMS'
+)
+@ddt.ddt
+class TestTpaHintSessionEndingDecorator(TestCase):
+
+    url = '/protected_view'
+
+    def setUp(self):
+        super(TestTpaHintSessionEndingDecorator, self).setUp()
+        self.enable_saml()
+        self.factory = RequestFactory()
+
+    def get_user_mock(self, authenticated=False):
+        user = MagicMock()
+        user.is_authenticated.return_value = authenticated
+        return user
+
+    def get_request_mock(self, authenticated=False, hinted=False, done=False):
+        user = self.get_user_mock(authenticated)
+        params = {}
+        if hinted:
+            params['tpa_hint'] = 'saml-realprovider'
+        if done:
+            params['session_cleared'] = 'yes'
+        url = '{}?{}'.format(self.url, urlencode(params))
+        request = self.factory.get(url)
+        request.user = user
+        return request
+
+    def create_provider(self, protected=False):
+        self.configure_saml_provider(
+            enabled=True,
+            name="Fancy real provider",
+            idp_slug="realprovider",
+            backend_name="tpa-saml",
+            drop_existing_session=protected,
+        )
+
+    @ddt.unpack
+    @ddt.data(
+        (True, True, False, False, False),
+        (True, True, True, False, True),
+        (True, True, True, True, False),
+        (True, True, False, True, False),
+        (False, True, True, False, False),
+        (True, False, True, False, False),
```

Comment:
Because we're testing the decorator rather than a specific view, we can't use `self.client()` without tying ourselves to testing the decorator in the context of a specific view it's attached to. Using the RequestFactory lets us avoid making our test implementation-specific; once we do the redirect, it's up to the view to which we redirect to ensure that, upon logout, the user is sent back to the correct place. We've tested to ensure that, in all of the states that we might encounter at the decorator, we do the right things - things which will not result in an infinite loop (e.g., we test both that when we redirect, we add the `session_cleared=yes` parameter to the return URL, and we test that when we receive a `session_cleared=yes` parameter, we pass through rather than redirect).